crypto/x509/x509_vfy.c

```diff
@@ -3224,11 +3223,11 @@ static int build_chain(X509_STORE_CTX *ctx)
             if (!ossl_assert(num == ctx->num_untrusted))
                 goto int_err;
             curr = sk_X509_value(ctx->chain, num - 1);
-            issuer = (self_signed || num > depth) ?
+            issuer = (X509_self_signed(curr, 0) || num > max_depth) ?
```

Comment:
> the intent is to not change the semantics

It is this change right here that bothers me.  The old code looked at the previous value of `self_signed` and this change is looking at the *current* cert.  Are you positive that the old `curr/self_signed` pair is unchanged?  How can you be?  And if you are not, how can you be sure that the semantics aren't changing?

Without at least one of those assurances, please do not put this code into 3.0
 
 To quote the running gag: "I've got a bad feeling about this."

Reply:
This indeed looks suspicious.
Yet note that further down in the same block the following happened:
```
            curr = issuer;
            if ((self_signed = X509_self_signed(curr, 0)) < 0)
                goto int_err;
```
That is, in the next iteration `curr` already points to the very same cert.
I've just tried to falsify this stipulation by running all tests with an equality check, and this succeeded. Which is of course no proof, but a pretty good reassurance.

Having a closer look at this I found that was rather bad coding style at the trailing sections if the major case distinctions in the loop body to re-assign `curr` because this misleads the reader that the value is needed for the next iteration, but in fact it is not needed!
The only case where it was needed was in the diagnostic code after loop termination, where it is much more clear to replace `curr` by `sk_X509_value(ctx->chain, num -1)`.

So thanks to your comment questioning the change, I've been able to even strengthen it by the commit I've just pushed.

Reply:
As I wrote, the `build_chain()` code is still somewhat messy,
but step by step we'll manage cleaning up this critical piece of code.

> To quote the running gag: "I've got a bad feeling about this."

To adapt that quote: "I've now got an even better feeling about this.".